charts/spinnaker/templates/configmap/halyard-config.yaml

```
@@ -90,10 +90,12 @@ data:
     {{ end }}
 
     # Docker Registry
-    $HAL_COMMAND config provider docker-registry enable
+    {{ $skipValidation := 0 }}
+    {{ range $index, $registry := .Values.dockerRegistries }}{{ if $registry.noValidate }}{{ $skipValidation = 1 }}{{ end }}{{ end }}
```

Can be shorter, but stayed consistent in code style:

```suggestion
    {{ range .Values.dockerRegistries }}{{ if .noValidate }}{{ $skipValidation = 1 }}{{ end }}{{ end }}
```

---

👍 Couldn't find a way to write it as a one-liner even with functions from https://helm.sh/docs/chart_template_guide/function_list/. I think the only thing we could consider is whether we should push the variable to the helpers template. Wdyt?

---

Putting it in the helpers template SGTM. Would rename the variable then. Couldn't come up with a good name, something like `shouldSkipGlobalDockerRegistryValidation`? 🤔

---

Hmm now that I am trying it, these helpers using `define` are more used like substituting an `{{ template "identifier" }}` with the contents of what is defined (e.g. [here](https://github.com/PicnicSupermarket/spinnaker-helm/blob/87eb09ec6b2fc096375247c37f913fdc77905ef5/charts/spinnaker/templates/configmap/halyard-config.yaml#L122-L128)). IIUC (didn't test) this would like like the following: 

```diff
diff --git a/charts/spinnaker/templates/_helpers.tpl b/charts/spinnaker/templates/_helpers.tpl
index c7324df..e8cd094 100644
--- a/charts/spinnaker/templates/_helpers.tpl
+++ b/charts/spinnaker/templates/_helpers.tpl
@@ -36,6 +36,16 @@ app: {{ include "spinnaker.fullname" . | quote }}
 release: {{ .Release.Name | quote }}
 {{- end -}}
 
+{{/*
+Create global Docker registry configuration validation skip variable
+iff there is at least one registry configured to skip this validation.
+*/}}
+{{- define "skipGlobalDockerRegistryValidation" -}}
+{{ $skipValidation := false }}
+{{ range $index, $registry := .Values.dockerRegistries }}{{ if $registry.noValidate }}{{ $skipValidation = 1 }}{{ end }}{{ end }}
+{{ if $skipValidation -}} --no-validate {{- end }}
+{{- end -}}
+
 {{/*
 Create comma separated list of namespaces in Kubernetes
 */}}
diff --git a/charts/spinnaker/templates/configmap/halyard-config.yaml b/charts/spinnaker/templates/configmap/halyard-config.yaml
index 223a96a..f5b5736 100644
--- a/charts/spinnaker/templates/configmap/halyard-config.yaml
+++ b/charts/spinnaker/templates/configmap/halyard-config.yaml
@@ -90,9 +90,7 @@ data:
     {{ end }}
 
     # Docker Registry
-    {{ $skipValidation := false }}
-    {{ range $index, $registry := .Values.dockerRegistries }}{{ if $registry.noValidate }}{{ $skipValidation = true }}{{ end }}{{ end }}
-    $HAL_COMMAND config provider docker-registry enable {{ if $skipValidation -}} --no-validate {{- end }}
+    $HAL_COMMAND config provider docker-registry enable {{ template "skipGlobalDockerRegistryValidation" }}
     {{- range $index, $registry := .Values.dockerRegistries }}
 
     if $HAL_COMMAND config provider docker-registry account get {{ $registry.name }} {{ if $registry.noValidate -}} --no-validate {{- end }}; then
```

WDYT?

---

I think that's one use case, but I also see e.g. within Picnic variable-like use case (see the Vault helpers).

So I was thinking of something like:

```diff
diff --git a/charts/spinnaker/templates/_helpers.tpl b/charts/spinnaker/templates/_helpers.tpl
index c7324df..c1e46c3 100644
--- a/charts/spinnaker/templates/_helpers.tpl
+++ b/charts/spinnaker/templates/_helpers.tpl
@@ -81,3 +81,7 @@ Create name of kubeconfig file to use when setting up kubernetes provider
 {{- printf "/opt/kube/%s" .Values.kubeConfig.secretKey  | toString -}}
 {{- end }}
 {{- end }}
+
+{{- define "skipDockerRegistry" -}}
+{{ range $index, $registry := .Values.dockerRegistries }}{{ if $registry.noValidate }}{{ $registry.noValidate }}{{ end }}{{ end }}
+{{- end }}
\ No newline at end of file
diff --git a/charts/spinnaker/templates/configmap/halyard-config.yaml b/charts/spinnaker/templates/configmap/halyard-config.yaml
index 34ed94d..5d79d10 100644
--- a/charts/spinnaker/templates/configmap/halyard-config.yaml
+++ b/charts/spinnaker/templates/configmap/halyard-config.yaml
@@ -90,9 +90,7 @@ data:
     {{ end }}
 
     # Docker Registry
-    {{ $skipValidation := 0 }}
-    {{ range $index, $registry := .Values.dockerRegistries }}{{ if $registry.noValidate }}{{ $skipValidation = 1 }}{{ end }}{{ end }}
-    $HAL_COMMAND config provider docker-registry enable {{ if $skipValidation -}} --no-validate {{- end }}
+    $HAL_COMMAND config provider docker-registry enable {{ if (include "skipDockerRegistry" .) -}} --no-validate {{- end }}
     {{- range $index, $registry := .Values.dockerRegistries }}
 
     if $HAL_COMMAND config provider docker-registry account get {{ $registry.name }} {{ if $registry.noValidate -}} --no-validate {{- end }}; then
```
which seems pretty clean and short. :) Note the use of `include` which allows "returning" values unlike `template` IIUC. WDYT?

---

As for the variable name, `skipGlobalDockerRegistryValidation` SGTM, but something like `skipDockerRegistryEnablingValidation` would be more precise, but no strong preference.

---

Nice suggestion. TIL about Helm's introduced `include` directive. Verified that this works by rebasing this on top of the `pdsoels/fix-deployment` branch and running `helmfile diff`. Also validated the flag is not set when `noValidate` is not set or set to false. Will push your suggestion with some other things.